src/Templates/Boilerplate/Bit.Boilerplate/src/Server/Boilerplate.Server.Api/Controllers/Identity/IdentityController.cs

```diff
@@ -211,24 +216,72 @@ public async Task<ActionResult<SignInResponseDto>> SignIn(SignInRequestDto reque
                 throw new ResourceValidationException(updateResult.Errors.Select(e => new LocalizedString(e.Code, e.Description)).ToArray());
         }
 
+        user.Sessions.Add(userSession);
+        var addUserSessionResult = await userManager.UpdateAsync(user);
+        if (addUserSessionResult.Succeeded is false)
+            throw new ResourceValidationException(addUserSessionResult.Errors.Select(e => new LocalizedString(e.Code, e.Description)).ToArray());
+
         return Empty;
     }
 
+    /// <summary>
+    /// Creates a user session and adds its ID to the access and refresh tokens, but only if the sign-in is successful.
+    /// </summary>
+    private UserSession GenerateUserSession()
+    {
+        var userAgentParseResult = DeviceDetector.GetInfoFromUserAgent(Request.Headers.UserAgent);
+
+        var device = userAgentParseResult.Success ?
+                                $"{userAgentParseResult.Match.BrowserFamily}, {userAgentParseResult.Match.OsFamily} {userAgentParseResult.Match.Os.Version}"
+                                : userAgentParseResult.ToString();
+
+        var userSession = new UserSession
+        {
+            SessionUniqueId = Guid.NewGuid(),
+            // Relying on Cloudflare cdn to retrieve address via IP
+            Address = string.Format("Country: {0}, Region: {1}, City: {2}, Postal Code: {3}, Lat: {4}, Lon: {5}",
+                Request.Headers["cf-ipcountry"], Request.Headers["cf-region"],
+                Request.Headers["cf-ipcity"], Request.Headers["cf-postal-code"],
+                Request.Headers["cf-iplatitude"], Request.Headers["cf-iplongitude"]),
+            Device = device,
+            IP = HttpContext.Connection.RemoteIpAddress?.ToString(),
+            StartedOn = DateTimeOffset.UtcNow
+        };
+
+        ((AppUserClaimsPrincipalFactory)userClaimsPrincipalFactory).SessionClaims.Add(new("session-id", userSession.SessionUniqueId.ToString()));
+
+        return userSession;
+    }
+
     [HttpPost]
     public async Task<ActionResult<TokenResponseDto>> Refresh(RefreshRequestDto request)
     {
         var refreshTokenProtector = bearerTokenOptions.Get(IdentityConstants.BearerScheme).RefreshTokenProtector;
         var refreshTicket = refreshTokenProtector.Unprotect(request.RefreshToken);
 
-        if (refreshTicket?.Properties?.ExpiresUtc is not { } expiresUtc || DateTimeOffset.UtcNow >= expiresUtc ||
-                await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not User user)
+        UserSession? userSession = null;
+
+        if (refreshTicket?.Properties?.ExpiresUtc is not { } expiresUtc ||
```

Comment:
if found user, but refresh token was expired, then remove session, if everything is fine, set sessions.RenewedOn, and in finally, save the user